### DISCUSSION TITLE
SSR window (global value) problem and must to refresh pages to fix it

### DISCUSSION THREAD

**Discussion by hpachy** [vercel/next.js#14980] [Help]
Hello! 

I have one question, I recently found something very weird with my code and I would like to know if it's possible to fix it.

My problem start with GTM implementation if you are curious I give you the SOF post => https://stackoverflow.com/questions/62791299/typeerror-window-gtag-is-not-a-function

The fact is that I discovered that's window global variable (and probably others I don't know) is equal to undefined because of the SSR so I catch error like : 
> TypeError: window.gtag is not a function

other example :
I would like to fill a form but at the submit my react state component obj doesn't have any value, he's totally empty and after a refresh this working like he must.
 

With more patience I found other bug on the chat at the bottom right of the website, ect... they does not appear at first! But when the app is refresh everything work fine, the chat come back and the other react component work well after, when I come back still work fine probably of the cache I guess...

So I would like to know if someone is familiar with that's kind of problem and if yes how should I fix it. 
I'm very puzzled, I add last thing I just ask here to know it's my hippothese is right.

Thx and have a good day :)

**Comment by jamesmosier** [comment]
You'll want to check for the existence of the `window` before using `window`.

For example:

```
if (typeof window !== 'undefined') {
  window.gtag("config", GA_TRACKING_ID, {
    page_location: url,
  });
}
```

**Reply by hpachy** [reply]
I did something like that's (is that's the same ?) : 
`
  if (window) {
    window.gtag("config", GA_TRACKING_ID, {
    page_location: url,
  });
}`

but I got the following error : 

`
Error from the trackerPageView =>  TypeError: window.gtag is not a function
    at _app.js:1
    at _app.js:1
    at commons.c57c1be722ad069a7405.js:1
    at Array.map (<anonymous>)
    at Object.emit (commons.c57c1be722ad069a7405.js:1)
    at commons.c57c1be722ad069a7405.js:1
`

so I wonder if I don't messed up.... Have you an explication of the general problem : 

_With more patience I found other bug on the chat at the bottom right of the website, ect... they does not appear at first! But when the app is refresh everything work fine, the chat come back and the other react component work well after, when I come back still work fine probably of the cache I guess..._

I really puzzled :/

**Reply by jamesmosier** [reply]
`if (window)` is not the same as `if (typeof window !== 'undefined')`. Checking for `typeof` is safe for non-browser environments. Additionally Next.js optimizes code it finds that is wrapped with `if (typeof window !== 'undefined')`.

As for your second problem I have no idea as I don't know what the code looks like.

**Reply by jamesmosier** [reply]
It looks like `gtag` is not something globally available by default. You have to set it up yourself according to this document: https://developers.google.com/analytics/devguides/collection/gtagjs

`gtag` is just a shortcut for `function gtag(){dataLayer.push(arguments);}`

**Reply by hpachy** [reply]
Ok I'll try, thx of the second problem I'll try to reproduce it and give you access to my repo ;) It's not a problem if I give it to you tomorrow? because to reproduce the bug it's gonna take me little time :/ 




_____________________________________________________________________________________________________

I place the response here because I placed it at the wrong place.

hpachy:

Hello! 

Sadly I'm not able to reproduce the bug I crossed but it's look like he disappear quite strange :/ I change my GTM implementation like you suggest to me  but still go the same error....

`
Error from the trackerPageView =>  TypeError: window.gtag is not a function
    NextJS 5
        value
        value
        emit
        emit
        value
`
My code look like this :

`export default function trackPageView(url) {`
`   try {`
  `   if (typeof window !== "undefined") {`
`       window.gtag("config", "GTM-****", {`
      `   page_location: url,`
    `   });`
  `   }`
`   } catch (error) {`
     `console.log("Error from the trackerPageView => ", error);v
   `}`
 `}`

**Reply by jamesmosier** [reply]
@hpachy you have to define a `gtag` function somewhere in order to use it:

```js
function gtag(){ window.dataLayer && window.dataLayer.push(arguments); }
// now you can use it
window.gtag()
```

**Reply by hpachy** [reply]
I added it in the same file as trackPageView() like you suggest to me but this the same error. I probably missing something. Sorry for the inconvenience and thx a lot for your time ! NextJs team is very awesome !

**Reply by jamesmosier** [reply]
Give this a try:

```js
function gtag() { 
  window.dataLayer && window.dataLayer.push(arguments); 
}

export default function trackPageView(url) {
  try {
    if (typeof window !== "undefined") {
      gtag("config", "GTM-****", {
        page_location: url,
      });
    }
  } catch (error) {
    console.log("Error from the trackerPageView => ", error); 
  } 
}
```

**Reply by hpachy** [reply]
Yep I tried exactly the same, so I probably miss understood  something during the setup of GTM... I gonna try to remake everything from the beginning  ;)

**Comment by MahdiTa97** [comment]
I set this condition before calling the **gtag** script :
```js
if (window && window.gtag) {
    window.gtag('config', GaTrackingId, {
      page_path: url,
    });
}
```
because some times there is no **gtag** in window

**Reply by JohnBDonner** [reply]
@MahdiTa97 This fixed it for me, but why does this occur? It makes me believe that some pages might miss analytics data for page views...

One hypothesis I have is that pages pre-rendered via the Next.js `<Link/>` tags don't have the `.gtag` function... though I'm not exactly sure why that'd be the case.
